drivers/syslog/syslog_stream.c

```diff
@@ -0,0 +1,303 @@
+ *   This is the low-level, single character, system logging interface
+ *   provided for the driver interface.
+ *
+ * Input Parameters:
+ *   channel    - Handle to syslog channel to be used.
+ *   ch         - The character to add to the SYSLOG (must be positive).
+ *
+ * Returned Value:
+ *   On success, the character is echoed back to the caller. A negated errno
+ *   value is returned on any failure.
+ *
+ ****************************************************************************/
+
+static int syslog_stream_putc(FAR struct syslog_channel_s *channel, int ch)
+{
+  FAR struct syslog_stream_s *chan =
+    (FAR struct syslog_stream_s *)channel;
+  FAR struct lib_outstream_s *stream =
+    (FAR struct lib_outstream_s *)chan->stream;
+  ssize_t nbytes;
+  uint8_t uch;
+  irqstate_t flags;
+
+  flags = enter_critical_section();
+  uch = (uint8_t)ch;
+  nbytes = stream->puts(stream, &uch, 1);
+  leave_critical_section(flags);
+
+  /* Check if the write was successful.  If not, nbytes will be
+   * a negated errno value.
+   */
+
+  if (nbytes < 0)
+    {
+      return (int)nbytes;
+    }
+
+  return ch;
+}
+
+/****************************************************************************
+ * Name: syslog_stream_force
+ *
+ * Description:
+ *   Force output in interrupt context.
+ *
+ * Input Parameters:
+ *   channel    - Handle to syslog channel to be used.
+ *   ch         - The character to add to the SYSLOG (must be positive).
+ *
+ * Returned Value:
+ *   On success, the character is echoed back to the caller.  A negated
+ *   errno value is returned on any failure.
+ *
+ ****************************************************************************/
+
+static int syslog_stream_force(FAR struct syslog_channel_s *channel, int ch)
+{
+  FAR struct syslog_stream_s *chan =
+    (FAR struct syslog_stream_s *)channel;
+  FAR struct lib_outstream_s *stream =
+    (FAR struct lib_outstream_s *)chan->stream;
+  ssize_t nbytes;
+  uint8_t uch;
+
+  uch = (uint8_t)ch;
+  nbytes = stream->puts(stream, &uch, 1);
+
+  /* Check if the write was successful.  If not, nbytes will be
+   * a negated errno value.
+   */
+
+  if (nbytes < 0)
+    {
+      return (int)nbytes;
+    }
+
+  return ch;
+}
+
+/****************************************************************************
+ * Name: syslog_stream_flush
+ *
+ * Description:
+ *   Flush any buffer data in the file system to media.
+ *
+ * Input Parameters:
+ *   channel    - Handle to syslog channel to be used.
+ *
+ * Returned Value:
+ *   Zero (OK) on success; a negated errno value is returned on any failure.
+ *
+ ****************************************************************************/
+
+static int syslog_stream_flush(FAR struct syslog_channel_s *channel)
+{
+  FAR struct syslog_stream_s *chan =
+    (FAR struct syslog_stream_s *)channel;
+  FAR struct lib_outstream_s *stream =
+    (FAR struct lib_outstream_s *)chan->stream;
+  irqstate_t flags;
+
+  flags = enter_critical_section();
```

Comment:
should we move the lock to the mtdoutstream?

Reply:
Moving the lock into stream requires lots of work, we have to modify every stream.
Especially for those streams  who has no protection around nput/nget.
So lets just keep the lock here.